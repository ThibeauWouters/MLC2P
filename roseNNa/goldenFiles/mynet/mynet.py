# Import packages
import torch
import torch.nn as nn
import sys
import os
import timeit
import numpy as np
import pathlib
import sys, getopt

# Filepath for storing the results
FILEPATH = "../goldenFiles/mynet/"
NAME  = "mynet"

# Options with sys.argv
opts, args = getopt.getopt(sys.argv[1:],"n")
produce = True
for opt, _ in opts:
    if opt == "-n":
        produce = False

"""Our network architecture"""
class NeuralNetwork(nn.Module):
    """
    Implements a two-layered neural network for the C2P conversion. Note that hence the number of layers is fixed
    for this NN subclass! The activation functions are sigmoids.
    """

    def __init__(self, h1: int = 600, h2: int = 200) -> None:
        """
        Initialize the neural network class.
        :param name: String that names this network, in order to recognize it later on.
        :param h1: Size (number of neurons) of the first hidden layer.
        :param h2: Size (number of neurons) of the second hidden layer.
        """
        # Call the super constructor first
        super(NeuralNetwork, self).__init__()

        # For convenience, save the sizes of the hidden layers as fields as well
        self.h1 = h1
        self.h2 = h2

        # Define the weights:
        self.linear1 = nn.Linear(3, h1)
        self.linear2 = nn.Linear(h1, h2)
        self.linear3 = nn.Linear(h2, 1)

    def forward(self, x):
        """
        Computes a forward step given the input x.
        :param x: Input for the neural network.
        :return: x: Output neural network
        """

        x = self.linear1(x)
        x = torch.sigmoid(x)
        x = self.linear2(x)
        x = torch.sigmoid(x)
        x = self.linear3(x)
        return x

    def load_parameters(self, state_dict):
        """
        Loads the parameters, as saved before, into this network's layers. Note: assumes 2-layer network, same size.
        Explanation: we changed the details of the neural network architecture definition to have named modules in order
        to better control the pruning methods.

        :param state_dict: A state_dict containing parameters.
        :return: None
        """

        # First layer
        stack_0_weight = state_dict['stack.0.weight']
        stack_0_bias = state_dict['stack.0.bias']
        self.linear1.weight = nn.parameter.Parameter(stack_0_weight)
        self.linear1.bias = nn.parameter.Parameter(stack_0_bias)

        # Second layer
        stack_2_weight = state_dict['stack.2.weight']
        stack_2_bias = state_dict['stack.2.bias']
        self.linear2.weight = nn.parameter.Parameter(stack_2_weight)
        self.linear2.bias = nn.parameter.Parameter(stack_2_bias)

        # Third layer
        stack_4_weight = state_dict['stack.4.weight']
        stack_4_bias = state_dict['stack.4.bias']
        self.linear3.weight = nn.parameter.Parameter(stack_4_weight)
        self.linear3.bias = nn.parameter.Parameter(stack_4_bias)

# Two auxiliary functions are needed to load the weights from the state dict into a fresh network
def get_hidden_sizes_from_state_dict(state_dict):
    """
    Finds the sizes of the two hidden layers of our 2-layer architecture given a state dict.
    :param state_dict: State dict of saved parameters
    :return: h1, size of first hidden layer, and h2, size of second hidden layer
    """
    h1 = np.shape(state_dict['linear1.bias'])[0]
    h2 = np.shape(state_dict['linear2.bias'])[0]

    return h1, h2

def create_nn(state_dict):
    """
    Create a NeuralNetwork object if given a dictionary of the weights, with correct sizes for hidden layers.
    :param state_dict: State dictionary containing the weights of the neural network
    :return:
    """
    h1, h2 = get_hidden_sizes_from_state_dict(state_dict)
    model = NeuralNetwork(h1=h1, h2=h2)
    model.load_state_dict(state_dict)

    return model


# Load the state dict that we saved here to have the right weights
state_dict = torch.load(FILEPATH + "state_dict.pth")
model = create_nn(state_dict)
model = model.float()
print("Successfully opened saved network:")
print(model)

# Run inference on the model to save it as ONNX
inp = torch.ones(1,3)
# Or run this test case:
inp = torch.Tensor([[10.20413115, 12.02658484, 22.13129693]])
print("Observed input")
print(inp)

if produce:
    with open("inputs.fpp",'w') as f:
        inputs = inp.flatten().tolist()
        inpShapeDict = {'inputs': list(inp.shape)}
        inpDict = {'inputs':inputs}
        f.write(f"""#:set inpShape = {inpShapeDict}""")
        f.write("\n")
        f.write(f"""#:set arrs = {inpDict}""")
        f.write("\n")
        f.write("a")

def stringer(mat):
    s = ""
    for elem in mat:
        s += str(elem) + " "
    return s.strip()

logits = model(inp)

with open(FILEPATH + NAME + ".txt", "w") as f:
    f.write(stringer(list(logits.shape)))
    f.write("\n")
    f.write(stringer(logits.flatten().tolist()))

# Print
print("Observed output")
print(logits.flatten().tolist())

torch.onnx.export(model,               # model being run
                  inp,                         # model input (or a tuple for multiple inputs)
                  FILEPATH + NAME + ".onnx",   # where to save the model (can be a file or file-like object)
                  export_params=True,        # store the trained parameter weights inside the model file
                  opset_version=10,          # the ONNX version to export the model to
                  do_constant_folding=True,  # whether to execute constant folding for optimization
                  input_names = ['input'],   # the model's input names
                  output_names = ['output'], # the model's output names
                  )

torch.onnx.export(model,               # model being run
                  inp,                         # model input (or a tuple for multiple inputs)
                  FILEPATH + NAME + "_weights.onnx",   # where to save the model (can be a file or file-like object)
                  export_params=True,        # store the trained parameter weights inside the model file
                  opset_version=10,          # the ONNX version to export the model to
                  do_constant_folding=False,  # whether to execute constant folding for optimization
                  input_names = ['input'],   # the model's input names
                  output_names = ['output'], # the model's output names
                  )
