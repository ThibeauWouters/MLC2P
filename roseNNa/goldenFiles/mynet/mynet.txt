1 1
12.866581916809082